projects/core/src/state/services/state-persistance.service.spec.ts

```diff
@@ -0,0 +1,179 @@
+} as Storage;
+
+const winRef = {
+  get nativeWindow(): Window {
+    return {} as Window;
+  },
+  get sessionStorage(): Storage {
+    return sessionStorageMock;
+  },
+  get localStorage(): Storage {
+    return localStorageMock;
+  },
+} as WindowRef;
+
+describe('StatePersistanceService', () => {
+  let service: StatePersistenceService;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [
+        contextServiceMapProvider,
+        SiteContextParamsService,
+        StatePersistenceService,
+        { provide: WindowRef, useValue: winRef },
+      ],
+    });
+
+    service = TestBed.inject(StatePersistenceService);
+
+    spyOn(sessionStorageMock, 'setItem').and.stub();
+    spyOn(localStorageMock, 'setItem').and.stub();
+  });
+
+  it('should inject service', () => {
+    expect(service).toBeTruthy();
+  });
+
+  describe('synWithStorage', () => {
+    it('should update storage on each state update', () => {
+      const state = new Subject<number>();
+
+      service.syncWithStorage({ key: 'test', state$: state.asObservable() });
+
+      state.next(5);
+      expect(localStorageMock.setItem).toHaveBeenCalledWith(
+        'spartacus⚿⚿test',
+        '5'
+      );
+
+      state.next(4);
+      expect(localStorageMock.setItem).toHaveBeenCalledWith(
+        'spartacus⚿⚿test',
+        '4'
+      );
+    });
+
+    it('should update specific state on each state update', () => {
+      const state = new Subject<number>();
+
+      service.syncWithStorage({
+        key: 'test',
+        state$: state.asObservable(),
+        context$: of(['electronics-spa', 'USD']),
+      });
+
+      state.next(5);
+      expect(sessionStorageMock.setItem).not.toHaveBeenCalled();
+      expect(localStorageMock.setItem).toHaveBeenCalledWith(
+        'spartacus⚿electronics-spa⚿USD⚿test',
+        '5'
+      );
+
+      state.next(4);
+      expect(localStorageMock.setItem).toHaveBeenCalledWith(
+        'spartacus⚿electronics-spa⚿USD⚿test',
+        '4'
+      );
+    });
+
+    it('should update session storage on each state update', () => {
+      const state = new Subject<number>();
+
+      service.syncWithStorage({
+        key: 'test',
+        state$: state.asObservable(),
+        storageType: StorageSyncType.SESSION_STORAGE,
+      });
+
+      state.next(5);
+      expect(localStorageMock.setItem).not.toHaveBeenCalled();
+      expect(sessionStorageMock.setItem).toHaveBeenCalledWith(
+        'spartacus⚿⚿test',
+        '5'
+      );
+
+      state.next(4);
+      expect(sessionStorageMock.setItem).toHaveBeenCalledWith(
+        'spartacus⚿⚿test',
+        '4'
+      );
+    });
+
+    it('should restore state on context not provided', () => {
+      spyOn(localStorageMock, 'getItem').and.returnValue('5');
+
+      const state = new Subject<number>();
+
+      service.syncWithStorage({
+        key: 'test',
+        state$: state.asObservable(),
+        onRead: res => expect(res).toEqual(5),
```

Comment:
This `expect` placed in the callback potentially can be omitted when the implementation of `syncWithStorage` is broken. And it will be omitted silently.

Instead, for example, we could assign a value to a local variable in the `onRead` callback. And add an `expect` for this variable in line 137.

(Same for similar cases in this file)